### PR TITLE
1214 sync emqx fork

### PR DESCRIPTION
**Pull request by zmstone**

### CODE REVIEW

**Review by ieQu1 — APPROVED** [review]

File: changelog.md [hunk 1]
```diff
@@ -2,6 +2,20 @@
    - Ported changes from EMQX's fork (based on 2.3.6) back to master branch
      - Included an pushback twards 'no_acl' callers. https://github.com/emqx/kafka_protocol/pull/1
    - Avoid crashing on decoding unknown error codes
+   - Improve SNI (server_name_indication) config.
+     - Prior to this change, SNI is auto-added only when SSL option
+       'verify' is set to 'verify_peer'.
+       This retriction is unnecessary, because SNI is a part of
+       client-hello in the handsake, it does not have anything to do
```
Comment (thalesmg) [inline changelog.md]:
```suggestion
       client-hello in the handshake, it does not have anything to do
```

File: src/kpro_connection.erl [hunk 2]
```diff
@@ -295,38 +295,39 @@ get_tcp_mod(_SslOpts = true)  -> ssl;
...
-
-insert_server_name_indication(_, SslOpts, _) ->
-  SslOpts.
+%% insert {server_name_indication, Host} if not already present
+%% Some special values are allowed:
+%% * auto: use the host as SNI
+%% * none: do not use SNI
+%% * undefined: same as auto (for backward compatibility)
+insert_server_name_indication(SslOpts0, Host) ->
+  SNI = ensure_string(proplists:get_value(server_name_indication, SslOpts0)),
```
Comment (thalesmg) [inline src/kpro_connection.erl:304]: Tiny nit: maybe we could ensure it is a string only if it's not one of the expected atoms (undefined, none, auto) to distinguish them from input?

**Review by thalesmg — APPROVED** [review]

**Review by thalesmg — COMMENTED** [review]
~~Should there be appup changes?~~

File: src/kpro_connection.erl (hunk 2 above)
Comment (zmstone) [reply src/kpro_connection.erl:304]: allowing in string format `"auto"` and `"none"` will make it easier for configs parsed from HOCON or JSON so there is no need to convert to atom.

**Pull request merged**